### DISCUSSION TITLE
Events raised in wrong browser when opening popup

### DISCUSSION THREAD

**Discussion by aimcosoftware** [cefsharp/CefSharp#4445] [General]
I have a `ChromiumWebBrowser` with handlers, including `DisplayHandler` attached.

calling `window.open()` creates a native window as expected, but handler events for the popup are sent to the opener. If the opener has a `LifespanHandler`, closing the native popup, raises `DoClose()` on it.

It doesn't raise `ChromiumWebBrowser` events, e.g. `Browser.FrameLoadStart()`, is not raised, only attached handler events.

Opening DevTools on the popup, opens the URL of the popup, not DevTools.

This also happens if you supply `newBrowser` in `OnBeforePopup`.

Parent browser initialization
```
Class Wrapper : Inherits Control
  Implements IDisplayHandler
  Private Browser As ChromiumWebBrowser()

  Sub Initialize(Url)
    Browser = New ChromiumWebBrowser(Url)
    Browser.LifeSpanHandler = Locals.LifeSpanHandler
    Browser.DisplayHandler = Me
    Browser.BrowserSettings = Settings()
    Browser.Dock = DockStyle.Fill
    Controls.Add(Browser)
  End Sub
...

End Class
```
LifespanHandler

```
Class LifeSpanHandler : Implements ILifeSpanHandler
  Function OnBeforePopup(WebBrowser As IWebBrowser, browser As IBrowser, frame As IFrame, targetUrl As String, targetFrameName As String, targetDisposition As WindowOpenDisposition, userGesture As Boolean, popupFeatures As IPopupFeatures, windowInfo As IWindowInfo, browserSettings As IBrowserSettings, ByRef noJavascriptAccess As Boolean, ByRef newBrowser As IWebBrowser) As Boolean Implements ILifeSpanHandler.OnBeforePopup
    Return False
  End Function
...
End Class
```

**Comment by amaitland** [comment]
> If the opener has a `LifespanHandler`, closing the native popup, raises `DoClose()` on it.

If I understand you correctly then this sounds like the `CEF` default behaviour to me. All `Popups` use the `Handlers` of their parent browser. You can use [IBrowser.IsPopup](https://cefsharp.github.io/api/110.0.x/html/P_CefSharp_IBrowser_IsPopup.htm) to determine if the handler method was called for a popup or the main browser.

Using `DoClose` as an example a very simple implementation might look like the following:

```c#
protected override bool DoClose(IWebBrowser chromiumWebBrowser, IBrowser browser)
{
	if (browser.IsPopup)
	{
		// default popups e.g. DevTools need to return false
		// so WM_CLOSE is sent to close the window.
		return false;
	}

	// Return true to cancel the close message being sent
	return true;
}
```

> Opening DevTools on the popup, opens the URL of the popup, not DevTools.

That should hopefully be fixed by #4443

**Reply by aimcosoftware** [reply]
> this sounds like the `CEF` default behaviour to me

Makes sense for native popups to raise events on opener.

> That should hopefully be fixed by https://github.com/cefsharp/CefSharp/pull/4443

It has for native popups, but not when returning newBrowser (but maybe it will when newBrowser is set correctly).

I'm still trying to create and return my own instance in newBrowser, and have it behave the same as one I have created.

Setting

**Reply by aimcosoftware** [reply]
@amaitland Apologies, I hadn't read the example you gave properly - I missed the setAsPopup step.

**Reply by aimcosoftware** [reply]
Thanks for your help on this, events now work as expected.

Now have something like this:
```basic
Function OnBeforePopup(...) As Boolean Implements ILifeSpanHandler.OnBeforePopup
  
  Browser = New ChromiumWebBrowser()
  Browser.SetAsPopup
  Browser.Dock = DockStyle.Fill
  Controls.Add(Browser)

  WindowInfo.SetAsChild(Browser.Handle)
  newBrowser = Browser

  Return False

End Function
```
Unfortunately the issue with `ShowDevTools` still applies when returning `newBrowser`. I'm using V111.2.20.0

**Reply by aimcosoftware** [reply]
When using `SetAsPopup`, would you expect `OnBeforePopup` to manage and route  `targetFrameName`?

**Reply by amaitland** [reply]
> Unfortunately the issue with `ShowDevTools` still applies when returning `newBrowser`. I'm using V111.2.20.0

PR https://github.com/cefsharp/CefSharp/pull/4443 was merged into `master`, it hasn't made it into an official release just yet. As it's a complex change my preference would be to include it in the next major release. `Chrome` is due to go stable with `M112` in today/tomorrow, so the next major release should be all going well within the next week.

Feedback welcome.

`M112` nightly test builds should be available later today.

> When using `SetAsPopup`, would you expect `OnBeforePopup` to manage and route `targetFrameName`?

I'm not quite sure what you mean by this. Where would `targetFrameName` be routed exactly?

**Reply by aimcosoftware** [reply]
Yes, not a problem for release suggestion, thanks

Should CEF/CEFSharp be sending subsequent calls to browsers with the same targetFrameName, or should I be doing that?

**Reply by amaitland** [reply]
Build https://www.myget.org/feed/cefsharp/package/nuget/CefSharp.WinForms/112.2.60-CI4749 is now available for testing.

There's an example [Nuget.config](https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/NuGet.config) that references the `MyGet` feed if required.

> Should CEF/CEFSharp be sending subsequent calls to browsers with the same targetFrameName, or should I be doing that?

I'm not sure. Do you happen to know how `Chrome` behaves?

**Reply by aimcosoftware** [reply]
@amaitland Great, thanks. Especially for the config link, Nuget is a mystery to me ;-)

I'll have play with targetFrameName and see what it needs.

**Reply by amaitland** [reply]
> Nuget is a mystery to me ;-)

You can put the `nuget.config` file next to your `sln` file as per https://learn.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior#config-file-locations-and-uses
 
You'll need to close and restart Visual Studio, then you should be able to install packages from the `MyGet` feed.

**Reply by aimcosoftware** [reply]
@amaitland OK, great thanks - I'm happy to Wiki for popups if that would be helpful.

**Reply by aimcosoftware** [reply]
@amaitland Sorry, won't download package, no real error either, all OK until:
```
Added package 'CefSharp.Common.112.2.60-CI4749' to folder 'C:\Users\DevWin10\Projects\_.NET Projects\Aimco DOMView\V111\packages' from source 'https://www.myget.org/F/cefsharp/'
Install failed. Rolling back...

```

**Reply by amaitland** [reply]
> I'm happy to Wiki for popups if that would be helpful.

That would be awesome 👍 Anyone with a `GitHub` account can add/edit the `Wiki`, feel free to create a new `wiki page` for popups and add anything you think might be useful for others 💯 

>  no real error either, all OK until:

I'm not seeing any issues on this end, I've upgraded the `MinimalExample` as a test. Installed without error, so I'm fairly sure the packages were built successfully.

I've pushed the changes to https://github.com/amaitland/CefSharp.MinimalExample/tree/beta/112 for reference.

Can you try using the `Package Manager Console` from within `Visual Studio`. If you use the `-Verbose` arg to get some more detail

Either of the following should work:
```
Install-Package CefSharp.WinForms -Version 112.2.60-CI4749 -Verbose

Install-Package CefSharp.WinForms -Version 112.2.60-CI4749 -Verbose -Source https://www.myget.org/F/cefsharp/api/v3/index.json
```

**Reply by aimcosoftware** [reply]
Very strange
```
Install failed. Rolling back...
Package 'CefSharp.Common.112.2.60-CI4749 : cef.redist.x64 [112.2.6, 112.2.6], cef.redist.x86 [112.2.6, 112.2.6]' does not exist in project 'Tab Form'
```
Tab Form is a project in my solution, but not sure why this is an issue. I'll just get it manually for now.

**Reply by amaitland** [reply]
> cef.redist.x64 [112.2.6, 112.2.6], cef.redist.x86 [112.2.6, 112.2.6]' does not exist in project

Double checking and both are on the `Nuget` feed.

- https://www.myget.org/feed/cefsharp/package/nuget/cef.redist.x64/112.2.6
- https://www.myget.org/feed/cefsharp/package/nuget/cef.redist.x86/112.2.6

I'll clear my package cache in a minute and perform a clean download.

**Reply by aimcosoftware** [reply]
They are downloaded. I've just tried a new project with the same result

**Reply by aimcosoftware** [reply]
Log
[nuget.txt](https://github.com/cefsharp/CefSharp/files/11153609/nuget.txt)

**Reply by amaitland** [reply]
I've cleared my local package cache, restored the solution without issue. Downloaded as expected, so I'm at a little bit of a loss on this one.

> nuget.txt

I've reviewed the log and cannot see a reason for the missing packages, I downloaded them manually from myget via the website just to be doubly sure, everything looks fine.

If you have a few mins are you able to checkout the https://github.com/amaitland/CefSharp.MinimalExample/tree/beta/112 branch and build CefSharp.MinimalExample.sln

```
git clone -b beta/112 https://github.com/amaitland/CefSharp.MinimalExample.git
```

**Reply by aimcosoftware** [reply]
> are you able to checkout...

Yes, that has downloaded the myget packages and runs.

**Reply by aimcosoftware** [reply]
@amaitland Sorry V112.2.60-CI4749 is not working as expected. definitely window handle issues , but I haven't had time to work it out yet.

**Reply by aimcosoftware** [reply]
Apart from the issue of `ShowDevTools`, showing the content of the calling page, everything else works as expected.

**Reply by aimcosoftware** [reply]
May have spoken too soon, I have changed my LifeSpanHandler to inherit from the Winforms version, and it may now be working.

**Reply by amaitland** [reply]
> May have spoken too soon, I have changed my LifeSpanHandler to inherit from the Winforms version, and it may now be working.

Sounds promising 👍 If you do experience any more issues if you could provide an example that reproduces the problem then I'll have a look.

**Reply by aimcosoftware** [reply]
> If you do experience any more issues

None so far, have tested this scenario:

``` c#
tab0.ShowDevTools() - (previously OK) - OK
tab0>devTools>window.open("url", "tab1") - (previously OK) - OK
tab1.ShowDevTools() - (previously opened "url") - OK

tab1>devTools>window.open("url", "popup1", "popup") - OK
popup1.ShowDevTools() - OK
popup1>devTools>window.open("url", "tab1") - OK
popup1>devTools>opener.close() - OK
popup1>devTools>close() - OK
```

**Reply by aimcosoftware** [reply]
Forgot to mention that when setting `WindowInfo.SetAsChild()` and returning `newBrowser`, CEF handles `targeFrameName`, so `OnbeforePopup()` is not raised when the target already exists.
